commcare_connect/program/api/serializers.py

```
@@ -0,0 +1,332 @@
+            raise serializers.ValidationError({"api_key": _("This API key is already registered to another user.")})
+
+    def _get_hq_app_name(self, api_key, domain, cc_app_id):
+        """Fetch the app name from HQ. Results cached per (api_key, domain) to avoid duplicate calls."""
+        if not hasattr(self, "_apps_cache"):
+            self._apps_cache = {}
+        cache_key = (api_key.pk, domain)
+        if cache_key not in self._apps_cache:
+            try:
+                self._apps_cache[cache_key] = get_applications_for_user_by_domain(api_key, domain)
+            except CommCareHQAPIException:
+                raise serializers.ValidationError({"non_field_errors": [_("Failed to fetch apps from CommCare HQ.")]})
```

Ruff (0.15.12)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)